### PR TITLE
Conversion tweaks

### PR DESCRIPTION
**Pull request by willrogers**
@TimGuiteDiamond a refactor of fromBob.

Also a few tweaks to make the conversion work better.

### CODE REVIEW

File: src/components/Widget/widget.tsx [hunk 1]
```diff
@@ -262,7 +262,7 @@ export const PVWidget = (props: PVWidgetComponent): JSX.Element => {
 
   const components = [];
 
-  if (props.actions) {
+  if (props.actions && props.actions.actions.length > 0) {
```
Comment (TimGuiteDiamond) [inline src/components/Widget/widget.tsx]: Any particular reason for this? Obviously a button with one action is not going to have *much* use for a MenuWrapper but perhaps the added description can be useful sometimes?

File: src/widgetActions.test.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+});
+
+describe("executeActions", (): void => {
+  it.each<[WidgetActions, number]>([
+    [ACTIONS_EX_AS_ONE, 2],
+    [ACTIONS_EX_FIRST, 3] // cumulative? we should reset it
+  ])("executes %d actions", (actions, num): void => {
+    executeActions(actions);
+    expect(writePv).toHaveBeenCalledTimes(num);
+  });
```
Comment (TimGuiteDiamond) [inline src/widgetActions.test.tsx]: Could split this into two tests to make it clearer what is happening?
- It executes multiple actions at once
- It executes one of multiple actions

File: src/App.tsx [hunk 3]
```diff
@@ -53,7 +53,7 @@ const App: React.FC = (): JSX.Element => {
   ];
   if (settings.coniqlSocket !== undefined) {
     const coniql = new ConiqlPlugin(settings.coniqlSocket);
-    plugins.push(["pva://", coniql]);
+    plugins.unshift(["pva://", coniql]);
```
[the remark this answers is not in the document]
Comment (willrogers) [reply src/App.tsx]: Just that the fallback mechanism relied on `""` being at the end of the list, so if you added pva to the end it didn't work. That could be more elegant but this fixed it.

File: src/components/FromBob/bobConversionUtils.ts [hunk 4]
```diff
@@ -17,110 +20,76 @@ interface BobDescription {
...
 
-export const bobVisibleToBoolen = (
-  inputProps: UnknownPropsObject,
-  outputProps: UnknownPropsObject
-): void => {
+export const bobParseBoolean = (
+  name: string,
+  jsonProp: convert.ElementCompact
+): boolean => {
+  let boolValue = true;
```
[the remark this answers is not in the document]
Comment (willrogers) [reply src/components/FromBob/bobConversionUtils.ts]: True, probably better to raise an error if it's neither true nor false.

File: src/components/FromBob/bobConversionUtils.ts [hunk 5]
```diff
@@ -17,110 +20,76 @@ interface BobDescription {
...
-    actionsToProcess = bobAction;
-  } else {
-    actionsToProcess = [bobAction];
+  if (Array.isArray(jsonProp.action)) {
+    actionsToProcess = jsonProp.action;
+  } else if (jsonProp.action !== undefined) {
+    actionsToProcess = [jsonProp.action];
   }
 
   // Object of available actions
```
[the remark this answers is not in the document]
Comment (willrogers) [reply src/components/FromBob/bobConversionUtils.ts]: Yes, this should come from xml.

File: src/components/FromBob/bobConversionUtils.ts [hunk 6]
```diff
@@ -247,11 +206,14 @@ export const convertBobToWidgetDescription = (
...
 
-  log.debug(compactJSON);
-
   // Add display to top of JSON to be processed
   // Assumes top level widget is always display - valid for XML files
   compactJSON.display._attributes = { type: "display" };
+  // We don't care about the position of the top-level display widget.
+  // We place it at 0,0 within its container.
+  compactJSON.display.x = { _text: "0" };
+  compactJSON.display.y = { _text: "0" };
```
[the remark this answers is not in the document]
Comment (willrogers) [reply src/components/FromBob/bobConversionUtils.ts]: It's not. This overwrites the top-level display position. That position is used in e.g. EDM to choose where the window opens in the monitor, but if you're embedding the bob screen in a div it just offsets it to the right and below where you actually wanted to put it.

File: src/components/Label/label.tsx [hunk 7]
```diff
@@ -11,7 +11,9 @@ import {
 
 const LabelProps = {
   text: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
-  visible: PropTypes.bool
+  visible: PropTypes.bool,
+  height: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (willrogers) [reply src/components/Label/label.tsx]: Maybe this isn't useful, I forget.

File: src/components/Positioning/positioning.tsx [hunk 8]
```diff
@@ -44,6 +45,15 @@ export function widgetDescriptionToComponent(
...
   } = widgetDescription;
 
+  let Component: React.FC<any>;
+  if (widgetDict.hasOwnProperty(type)) {
+    log.warn(`Failed to load unknown widget type ${type}`);
+    Component = widgetDict[type];
+  } else {
+    Component = Shape;
+    backgroundColor = "magenta";
+  }
```
[the remark this answers is not in the document]
Comment (willrogers) [reply src/components/Positioning/positioning.tsx]: That log line is still in the wrong place. I thought I had fixed it.

File: src/components/Widget/widget.tsx (hunk 1 above)
Comment (willrogers) [reply src/components/Widget/widget.tsx]: Greater than zero, not greater than one.

File: src/widgetActions.test.tsx (hunk 2 above)
Comment (willrogers) [reply src/widgetActions.test.tsx]: Yes.

File: src/components/FromBob/bobConversionUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: EDM strikes again :laughing: 

Fair enough, this makes sense as we are likely to always want the 0 offset within the div we have specified the position for.

File: src/components/Positioning/positioning.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/Positioning/positioning.tsx]: Ah so it is, I just meant that catching and logging this error will be quite useful :)

File: src/components/Widget/widget.tsx (hunk 1 above)
Comment (TimGuiteDiamond) [reply src/components/Widget/widget.tsx]: So this prevents empty menu wrappers from being shown?

File: src/components/FromBob/bobConversionUtils.ts [hunk 9]
```diff
@@ -17,110 +20,76 @@ interface BobDescription {
...
-  }
+export const bobParseMacros = (
+  name: string,
+  jsonProp: convert.ElementCompact
+): MacroMap => {
+  const macroMap: MacroMap = {};
+  Object.entries(jsonProp as object).forEach(([key, value]): void => {
+    macroMap[key] = value["_text"];
+  });
+  return macroMap;
```
[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: Fixed in 71e3e79

**Comment by willrogers** [comment]
Yes, that's the main question. There's also the converse: what if multiple properties in a different representation want to be one property in ours.

I think we could have additional collection and separation functions to handle this once the initial conversion has taken place.

If we want to treat bob as a fully-supported file format, there are other questions too. One is: you find the variable `${pv_name}` (and other properties) used in other properties e.g. but not limited to tooltips. We would either have to adopt the same property names or have a more careful replacement mechanism to catch all of these. Another: 'exit' is not a supported action, so it's handled in scripts, at least in cs-studio. I wonder if we should propose it.

Work on this was inevitably leading me in the direction of defining widget property types more carefully. This might be the next piece of work to do, but it won't produce any more functionality, just clean up the internals.

File: src/components/FromBob/bobConversionUtils.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: It comes in here:
```xml
    <actions execute_as_one="true">
      <action type="write_pv">
        <pv_name>loc://pv1</pv_name>
        <value>50</value>
        <description>Write pv1 to 50</description>
      </action>
      <action type="write_pv">
        <pv_name>loc://pv2</pv_name>
        <value>Testing</value>
        <description>Write pv2 to "Testing"</description>
      </action>
    </actions>
```
in the `attributes`. Will add now.

[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: See #126

[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: Ready to merge #126

File: src/components/FromBob/bobConversionUtils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: Might be worth handling this before merging

[the remark this answers is not in the document]
Comment (willrogers) [reply src/components/FromBob/bobConversionUtils.ts]: as we speak

[the remark this answers is not in the document]
Comment (TimGuiteDiamond) [reply src/components/FromBob/bobConversionUtils.ts]: Looks good, should be ready to merge?

**Pull request merged**